http/README.md

```diff
@@ -0,0 +1,153 @@
+
+## HTTP Transport
+
+Nodes MUST use HTTPS (i.e., they MUST NOT use plaintext HTTP). It is RECOMMENDED to use HTTP/2 and HTTP/3.
+
+Nodes signal support for their HTTP transport using the `/http` component in
+their multiaddr. E.g., `/dns4/example.com/tls/http`. See the [HTTP multiaddr
+component spec](https://github.com/libp2p/specs/pull/550) for more details.
+
+## Namespace
+
+libp2p does not squat the global namespace. libp2p application protocols can be discovered by the [well-known resource](https://www.rfc-editor.org/rfc/rfc8615) `.well-known/libp2p`. This allows server operators to dynamically change the URLs of the application protocols offered, and not hard-code any assumptions how a certain resource is meant to be interpreted.
+
+```json
+
+{
+    "protocols": {
+        "/kad/1.0.0": {"path": "/kademlia/"},
+        "/ipfs-http/1.0.0": {"path": "/"},
+    }
+}
+```
+
+The resource contains a mapping of application protocols to their respective URL. For example, this configuration file would tell a client
+
+1. That the Kademlia protocol is available at `/kademlia` and
+2. The [IPFS Trustless Gateway API](https://specs.ipfs.tech/http-gateways/trustless-gateway/) is mounted at `/`.
+
+It is valid to expose a service at `/`. It is RECOMMENDED that implementations facilitate the coexistence of different service endpoints by ensuring that more specific URLs are resolved before less specific ones. For example, when registering handlers, more specific paths like `/kademlia/foo` should take precedence over less specific handler, such as `/`.
+
+## Peer ID Authentication
+
+When using the HTTP Transport, peer id authentication is optional. You only pay for it if you need it. This benefits use cases that don’t need peer authentication (e.g., fetching content addressed data) or authenticate some other way (not tied to libp2p peer ids).
+
+Peer ID authentication in the HTTP Transport follows a similar to pattern to how
+libp2p adds Peer ID authentication in WebTransport and WebRTC. We run the
+standard libp2p Noise handshake, but using `IX` for client and server
+authentication or `NX` for just server authentication.
+
+Note: This is just one form of Peer ID authentication. Other forms may be added
+in the future (with a different `WWW-Authenticate` value) or be added to the
+application protocols themselves.
+
+### Authentication flow
+
+1. The client initiates a request that it knows must be authenticated OR the client responds to a `401` with the header `WWW-Authenticate: Libp2p-Noise` (The server MAY also include `Libp2p-Token` as an authentication scheme).
+2. The client sets the `Authorization` [header](https://www.rfc-editor.org/rfc/rfc9110.html#section-11.6.2) to `Libp2p-Noise <multibase-encoded-noise-protobuf>` . This initiates the `IX` or `NX` handshake.
```

Comment:
We might run into the HTTP header size limit here (e.g. for HTTP/3: https://datatracker.ietf.org/doc/html/rfc9114#section-4.2.2), especially when using PQ keys.

Would it make sense to use a separate HTTP endpoint here, and do the handshake using the body of a request / response to that endpoint?

Reply:
How big are pq keys? It's pretty nice and ergonomic to do auth in the headers because then standard stuff like this works:

```go
func(w http.ResponseWriter, r *http.Request) {
    // This handler requires libp2p auth
    clientID, err := AuthenticateClient(serverKey, w.Header(), r.Header)
    
    if err != nil || clientID == "" {
	w.Header().Set("WWW-Authenticate", "Libp2p-Noise-IX")
	w.WriteHeader(http.StatusUnauthorized)
	return
    }
    
    w.WriteHeader(http.StatusOK)
    w.Write([]byte(fmt.Sprintf("Hello, %s", clientID)))
}
```

Or even middlewares:

```go
libp2pAuthMiddleware(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
	w.WriteHeader(http.StatusOK)
	w.Write([]byte(fmt.Sprintf("Hello, %s", r.Context().Value(libp2pAuthKey))))
}))
```

A [survey](https://www.geekersdigest.com/max-http-request-header-size-server-comparison/) shows that many support at least 8KB in headers.

We could only support the libp2p-token in the auth header, and do the the noise handshake on a dedicated endpoint. That would mean we wouldn't be able to piggyback on a request to do auth. It would also mean that doing something with client auth is always a 2 step process. It also adds a bit of complexity I think since there are two flows for auth, instead of everything happening in the headers.

If we can stay below the 8KB limit, [which may be possible](https://blog.cloudflare.com/sizing-up-post-quantum-signatures/) I think sticking with the headers for everything would be nicer.